Fix Interest stream should drop message when no interest based on FilterSubjects

MauriceVanVeen opened the pull request:
Resolves https://github.com/nats-io/nats-server/issues/5383

Signed-off-by: Maurice van Veen <github@mauricevanveen.com>

MauriceVanVeen commented:
`TestFileStoreFSSMeta` and `TestJetStreamClusterStreamCatchupNoState` fail in Travis, might they be flaky they pass locally?

Jarema commented:
Thanks for the PR!

Yes, the failing tests seems to be flaky ones.
Rerun then and its green now.

derekcollison commented on server/stream.go:
```diff
@@ -4716,10 +4716,19 @@ func (mset *stream) processJetStreamMsg(subject, reply string, hdr, msg []byte,
 			// Assume no interest and check to disqualify.
 			noInterest = true
 			mset.clsMu.RLock()
+		out:
```
This could have an impact if large number of consumers or large number of filter subjects. We already have these leaded into a sublist, so let's use that.

```go
noInterest := !mset.csl.hasInterest(subject)
```

MauriceVanVeen replied:
That nicely simplifies it, done!

derekcollison commented on server/stream.go:
```diff
@@ -4713,25 +4713,7 @@ func (mset *stream) processJetStreamMsg(subject, reply string, hdr, msg []byte,
-				}
-			}
-			mset.clsMu.RUnlock()
+			noInterest = !mset.csl.HasInterest(subject)
```
Maybe this? Even simpler?

```go
noInterest := numConsumers == 0 || !mset.csl.HasInterest(subject)
```

derekcollison replied:
Maybe check that cls is not nil here too.

MauriceVanVeen replied:
Done

MauriceVanVeen replied:
Have added the `mset.csl == nil` check as well.

Wondering whether it should also be under the `mset.clsMu` lock when checking?

derekcollison replied:
Yes it should be..

MauriceVanVeen commented:
Seems `TestJetStreamClusterInterestRetentionWithFilteredConsumers` fails since `mset.csl == nil`

MauriceVanVeen commented:
Seems `TestJetStreamClusterInterestRetentionWithFilteredConsumers` genuinely fails now.

Not sure why but `mset.csl` is sometimes `nil`, sometimes contains partial data. Also after re-adding the use of `mset.clsMu`.

derekcollison commented:
Hmm. I will take a look over weekend.

MauriceVanVeen commented:
Pushed a change that passes the failing tests locally.

When debugging I noticed that `mset.cList` would always be consistent, whereas `mset.csl` would not be.

To ensure consistency; everywhere `mset.cList` is changed, `mset.csl` is now also updated making them aligned. 
- `setConsumer` appends to the `cList` and also updates the `csl` now. (which was previously only done in `setConsumerAsLeader`.
- `removeConsumer` already updated the `cList` and also removed the `signalSubs` from the `csl` (which was previously also done in `removeConsumerAsLeader`)

Not entirely sure what caused the desync, but I did notice that the `TestJetStreamClusterInterestRetentionWithFilteredConsumers` did trigger a leader change at some point. So I'm assuming the `mset.csl` would only be set initially, but not after a leader change. Having `cList` and `csl` aligned should guard against leader changes as well.

derekcollison reviewed (changes requested):
When a stream uses csl to signal a consumer for message delivery, it should only be doing so for the consumers that are leaders since they are the only ones that can deliver messages. 

This change would have the stream notify all consumers even if they are not the leader.

derekcollison commented:
We can keep the change, but we need to track down `mset.signalConsumers()` and the underlying sub.icb which calls into `o.processStreamSignal`.

We acquire the lock in that code, so we need to add in a check for if we are the leader, and if not simply return. So in that code where we check for `mset == nil` also add in leader check.

```go
	o.mu.Lock()
	defer o.mu.Unlock()
	if o.mset == nil || !o.isLeader() {
		return
	}
```

derekcollison commented:
Also need to resolve conflict with main, so rebase probably needed.

derekcollison commented:
We are close, really nice work! Just need to tidy up signaling non-leaders now since we always populate csl.

MauriceVanVeen commented:
Have rebased onto top of main, and added the leader check in `o.processStreamSignal` here: https://github.com/nats-io/nats-server/pull/5384/commits/34574121e53534a700554e318ebcc23b10643bbd

derekcollison reviewed (approved):
LGTM - Thanks Maurice!
